hcl2template/types.required_plugins.go

```diff
@@ -41,24 +42,126 @@ func (cfg *PackerConfig) decodeRequiredPluginsBlock(f *hcl.File) hcl.Diagnostics
+	// plugin usages.
+
+	// decodeRequiredPluginsBlock needs to be called before
+	// decodeImplicitRequiredPluginsBlocks; otherwise all required plugins will
+	// be implicitly required too.
 
 	var diags hcl.Diagnostics
 
 	content, moreDiags := f.Body.Content(configSchema)
 	diags = append(diags, moreDiags...)
 
 	for _, block := range content.Blocks {
+
 		switch block.Type {
 		case sourceLabel:
-			// TODO
+			diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Builder, block)...)
+		case dataSourceLabel:
+			diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Datasource, block)...)
+		case buildLabel:
+			content, _, moreDiags := block.Body.PartialContent(buildSchema)
+			diags = append(diags, moreDiags...)
+			for _, block := range content.Blocks {
+
+				switch block.Type {
+				case buildProvisionerLabel:
+					diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Provisioner, block)...)
+				case buildPostProcessorLabel:
+					diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(PostProcessor, block)...)
+				case buildPostProcessorsLabel:
+					content, _, moreDiags := block.Body.PartialContent(postProcessorsSchema)
+					diags = append(diags, moreDiags...)
+					for _, block := range content.Blocks {
+
+						switch block.Type {
+						case buildPostProcessorLabel:
+							diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(PostProcessor, block)...)
+						}
+					}
+				}
+			}
+
 		}
 	}
 	return diags
 }
 
+func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlock(k ComponentKind, block *hcl.Block) hcl.Diagnostics {
+	if len(block.Labels) == 0 {
+		// malformed block ? Let's not panic :)
+		return nil
+	}
+	// Currently all block types are `type "component-kind" ["name"] {`
+	// this makes this simple.
+	componentName := block.Labels[0]
+
+	redirect := map[ComponentKind]map[string]string{
+		Builder:       cfg.parser.PluginConfig.BuilderRedirects,
+		PostProcessor: cfg.parser.PluginConfig.PostProcessorRedirects,
+		Provisioner:   cfg.parser.PluginConfig.ProvisionerRedirects,
+		Datasource:    cfg.parser.PluginConfig.DatasourceRedirects,
+	}[k][componentName]
+
+	if redirect == "" {
+		// no known redirect for this component
+		return nil
+	}
+
+	redirectAddr, diags := addrs.ParsePluginSourceString(redirect)
+	if diags.HasErrors() {
+		// This should never happen, since the map is manually filled.
+		return diags
+	}
+
+	for _, req := range cfg.Packer.RequiredPlugins {
+		if _, found := req.RequiredPlugins[redirectAddr.Type]; found {
+			// This could happen if a plugin was forked. For example, I forked
+			// the github.com/hashicorp/happycloud plugin into
+			// github.com/azr/happycloud that is required in my config file; and
+			// am using the `happycloud-uploader` pp component from it. In that
+			// case - and to avoid miss-requires - we won't implicitly import
+			// any other `happycloud` plugin.
+			return nil
+		}
+	}
+
+	store := map[ComponentKind]packer.BasicStore{
```

Comment:
Wouldn't be best to check this on top of everything? Even before checking if a redirect is available?

Reply:
Yup, moving that one